nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,492 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  isPath = x: builtins.substring 0 1 x == "/";
+ 
+  mkExcludeFile = cfg:
+    # Write each exclude pattern to a new line
+    pkgs.writeText "excludefile" (concatStringsSep "\n" cfg.exclude);
+
+  mkPathsArgs = cfg:
+    # Shell-escape each path and put spaces between them
+    concatMapStringsSep " " escapeShellArg cfg.paths;
+
+  mkKeepArgs = cfg:
+    # If cfg.prune.keep e.g. has a yearly attribute,
+    # its content is passed on as --keep-yearly
+    concatStringsSep " "
+      (mapAttrsToList (x: y: "--keep-${x}=${toString y}") cfg.prune.keep);
+
+  mkBackupScript = cfg: ''
+    on_exit()
+    {
+      exit_status=$?
+      # Reset the EXIT handler, or else we're called again on 'exit' below
+      trap - EXIT
+      ${cfg.postHook}
+      exit $exit_status
+    }
+    trap 'on_exit' INT TERM QUIT EXIT
+
+    ${cfg.preHook}
+  '' + optionalString cfg.doInit ''
+    # Run borg init if the repo doesn't exist yet
+    if ! borg list; then
+      ${optionalString (isPath cfg.repo) "mkdir -p ${escapeShellArg cfg.repo}"}
+      borg init \
+        --encryption ${cfg.encryption.mode} \
+        ${cfg.extraInitArgs}
+      ${cfg.postInit}
+    fi
+  '' + ''
+    borg create \
+      --compression ${cfg.compression} \
+      --exclude-from ${mkExcludeFile cfg} \
+      ${cfg.extraCreateArgs} \
+      "::$archiveName" \
+      ${mkPathsArgs cfg}
+    ${cfg.postCreate}
+  '' + optionalString (cfg.prune.keep != { }) ''
+    borg prune \
+      ${mkKeepArgs cfg} \
+      ${cfg.extraPruneArgs}
+    ${cfg.postPrune}
+  '' + ''
+    ${cfg.postHook}
+  '';
+
+  mkPassEnv = cfg: with cfg.encryption;
+    if passCommand != null then
+      { BORG_PASSCOMMAND = passCommand; }
+    else if passphrase != null then
+      { BORG_PASSPHRASE = passphrase; }
+    else { };
+
+  mkBackupService = name: cfg: 
+    let
+      userHome = config.users.users.${cfg.user}.home;
+    in nameValuePair "borgbackup-job-${name}" {
+      description = "BorgBackup job ${name}";
+      path = with pkgs; [
+        borgbackup openssh
+      ];
+      script = mkBackupScript cfg;
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        # Only run when no other process is using CPU or disk
+        CPUSchedulingPolicy = "idle";
+        IOSchedulingClass = "idle";
+        ProtectSystem = "full";
+        #FIXME:
```

Comment:
Perhaps do something about this FIXME before merge?

Reply:
The things necessary to use `ProtectSystem = "strict"`:
- assert that `createHome == true` (or don't assert it and just try to create the directory)
- create `$HOME/{.config,.cache}/borg`

Reply:
I went for the assertion, but `root` doesn't have an entry in `config.users` by default, so it is treated specially.

Reply:
Dropped the assertion again. Home dirs can also be created in other ways: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/users-groups.nix#L530-L533
I think the error messages should be clear enough though.